src/core/qobj.jl

```diff
@@ -119,12 +119,51 @@ struct Operator{T<:Complex}
     data::Matrix{T}
 end
 
-# overload constructor to enable initilization from Real-valued Matrix
+# overload constructor to enable initialization from Real-valued Matrix
 Operator(x::Matrix{T}) where {T<:Real} = Operator(convert(Matrix{Complex{T}},x) )
 
-# Constructor using Adjoint(Operator{T})
+# Constructor using adjoint(Operator{T})
 Operator(x::LinearAlgebra.Adjoint{T,Matrix{T}}) where {T<:Complex} = Operator{T}(x) 
 
+abstract type AbstractOperator end
+
+"""
+A structure representing a diagonal quantum operator (i.e. a complex matrix, with non-zero elements all lying on the diagonal).
+
+# Fields
+- `data` -- a vector containing the diagonal.
+
+# Examples
+```jldoctest
+julia> z = Snowflake.DiagonalOperator([1.0,-1.0])
+(2,2)-element Snowflake.DiagonalOperator:
+Underlying data type: ComplexF64:
+1.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    -1.0 + 0.0im
+
+julia> z = Snowflake.DiagonalOperator([1.0+im,1.0,1.0,0.0-im])
+(4,4)-element Snowflake.DiagonalOperator:
+Underlying data type: ComplexF64:
+1.0 + 1.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 - 1.0im
+
+```
+"""
+struct DiagonalOperator{N,T<:Complex}<:AbstractOperator
+    data::SVector{N,T}
```

Comment:
What is the performance difference when we use a vector instead of a static vector?